src/globus_sdk/experimental/tokenstorage_v2/base.py

```diff
@@ -0,0 +1,129 @@
+from __future__ import annotations
+
+import abc
+import contextlib
+import os
+import typing as t
+
+from globus_sdk.services.auth import OAuthTokenResponse
+
+
+class StorageAdapterV2(metaclass=abc.ABCMeta):
+    """
+    Abstract base class for interacting with an underlying storage system to manage
+    storage of token data.
+
+    The ``namespace`` is a user-supplied way of partitioning data, and any token
+    responses passed to the storage adapter are broken apart and stored indexed by
+    *resource_server*. If you have a more complex use-case in which this scheme will be
+    insufficient, you should encode that in your choice of ``namespace`` values.
+
+    If you have a more complex use-case in which this scheme will be
+    insufficient, you should encode that in your choice of ``namespace`` values.
+    """
+
+    def __init__(self, namespace: str = "DEFAULT") -> None:
+        """
+        :param namespace: A user-supplied namespace for partitioning token data.
+        """
+        self.namespace = namespace
+
+    @abc.abstractmethod
+    def store(self, token_response: OAuthTokenResponse | dict[str, t.Any]) -> None:
+        """
+        Store token data in the underlying storage under the current namespace.
+
+        :param token_response: Either an ``OAuthTokenResponse`` or a dict containing
+            in the same format as ``OAuthTokenResponse.by_resource_server``
+        """
+
+    @abc.abstractmethod
+    def get_token_data(self, resource_server: str) -> dict[str, t.Any] | None:
+        """
+        Lookup token data for a resource server in the underlying storage
+        under the current namespace.
+
+        Either returns a dict with the access token, refresh token (optional), and
+        expiration time, or returns ``None``, indicating that there was no data for that
+        resource server.
+
+        :param resource_server: The resource_server string to get tokens for
+        """
+
+    @abc.abstractmethod
+    def remove_tokens_for_resource_server(self, resource_server: str) -> bool:
+        """
+        Remove all tokens for a resource server from the underlying storage under the
+        current namespace.
+
+        Returns True if token data was deleted, False if none was found to delete.
+
+        :param resource_server: The resource server string to remove tokens for
+        """
+
+    def normalize_token_response(
```

Comment:
This doesn't feel like it should be exposed on the public interface for every storage adapter.

It's a convenient thing for implementing adapters to have access to, but maybe as `_normalize_token_response` or even as a separate method which can be imported?

Reply:
This has been removed